src/Npgsql/NpgsqlConnector.Auth.cs

```diff
@@ -67,19 +69,89 @@ async Task AuthenticateCleartext(string username, bool async)
+            var supportsSha256 = mechanisms.Contains("SCRAM-SHA-256");
+            var supportsSha256Plus = mechanisms.Contains("SCRAM-SHA-256-PLUS");
+            if (!supportsSha256 && !supportsSha256Plus)
+                throw new NpgsqlException("No supported SASL mechanism found (only SCRAM-SHA-256 and SCRAM-SHA-256-PLUS are supported for now). " +
                                           "Mechanisms received from server: " + string.Join(", ", mechanisms));
-            var mechanism = "SCRAM-SHA-256";
+
+            var mechanism = string.Empty;
+            var cbindFlag = string.Empty;
+            var cbind = string.Empty;
+            var successfulBind = false;
+
+            if (supportsSha256Plus)
+            {
+                // RFC 5929
+                mechanism = "SCRAM-SHA-256-PLUS";
+                // PostgreSQL only supports tls-server-end-point binding
+                cbindFlag = "p=tls-server-end-point";
+                // SCRAM-SHA-256-PLUS depends on using ssl stream, so it's fine
+                var sslStream = (SslStream)_stream;
+                var cbindFlagBytes = Encoding.UTF8.GetBytes($"{cbindFlag},,");
+
+                using var remoteCertificate = new X509Certificate2(sslStream.RemoteCertificate);
+                // Checking for hashing algorithms
+                HashAlgorithm? hashAlgorithm = null;
+                var algorithmName = remoteCertificate.SignatureAlgorithm.FriendlyName;
+                if (algorithmName.IndexOf("sha1", StringComparison.OrdinalIgnoreCase) >= 0 ||
+                    algorithmName.IndexOf("md5", StringComparison.OrdinalIgnoreCase) >= 0 ||
+                    algorithmName.IndexOf("sha256", StringComparison.OrdinalIgnoreCase) >= 0)
+                {
+                    hashAlgorithm = SHA256.Create();
+                }
+                else if (algorithmName.IndexOf("sha384", StringComparison.OrdinalIgnoreCase) >= 0)
+                {
+                    hashAlgorithm = SHA384.Create();
+                }
+                else if (algorithmName.IndexOf("sha512", StringComparison.OrdinalIgnoreCase) >= 0)
+                {
+                    hashAlgorithm = SHA512.Create();
+                }
+                else
+                {
+                    Log.Warn($"Support for signature algorithm {algorithmName} is not yet implemented, falling back to SCRAM-SHA-256");
+                }
+
+                if (hashAlgorithm != null)
+                {
+                    using (hashAlgorithm)
+                    {
+                        var certificateHash = hashAlgorithm.ComputeHash(remoteCertificate.GetRawCertData());
+                        var cbindBytes = cbindFlagBytes.Concat(certificateHash).ToArray();
+                        cbind = Convert.ToBase64String(cbindBytes);
+                        successfulBind = true;
+                        IsScramPlus = true;
+                    }
```

Comment:
Just meant this, to reduce a block. It really is a minor nitpick, am OK with not doing it.

```suggestion
                    using var _ = hashAlgorithm;
                    var certificateHash = hashAlgorithm.ComputeHash(remoteCertificate.GetRawCertData());
                    var cbindBytes = cbindFlagBytes.Concat(certificateHash).ToArray();
                    cbind = Convert.ToBase64String(cbindBytes);
                    successfulBind = true;
                    IsScramPlus = true;
```